### PR TITLE
test/system: Don't check the output of 'toolbox run'

### PR DESCRIPTION
**Pull request by HarryMichal**
Now that the rewritten version is the [default](#437), the tests need to be adjusted.

Also, we did not enable tests on Fedora 32 -> enabling with this.

### CODE REVIEW

**Comment by HarryMichal** [comment]
Looks like I'll need to adjust the parts where the tests check what was written in stdout/stderr. The spinner breaks those test cases.

**Comment by debarshiray** [comment]
> Looks like I'll need to adjust the parts where the tests check what was written
> in stdout/stderr. The spinner breaks those test cases.

Oh, really? Is it because the spinner just looks different now? I think that the spinner is written to the standard output in both the POSIX shell and Go versions.

File: playbooks/shellcheck.yaml [hunk 1]
```diff
@@ -5,7 +5,8 @@
       become: yes
       package:
         name:
-          - golang-github-cpuguy83-go-md2man
```
Comment (debarshiray) [inline playbooks/shellcheck.yaml:8]: It would be nicer to split the changes related to switching `playbooks/shellcheck.yaml` to Fedora 32 into a separate commit. For a moment, I was thinking that this change from `golang-github-cpuguy83-go-md2man` to `golang-github-cpuguy83-md2man` is related to some sort of typo.

File: test/system/102-create.bats [hunk 2]
```diff
@@ -7,14 +7,14 @@ load helpers
 }
 
 @test "Create a container with a valid custom name ('not-running')" {
-  run_toolbox -y create -c "not-running"
+  run_toolbox -y create "not-running"
```
Comment (debarshiray) [inline test/system/102-create.bats:10]: We still support the `--container` flag, so we should also continue to test it. :)

Comment (HarryMichal) [reply test/system/102-create.bats:10]: Ah, right.

File: playbooks/shellcheck.yaml (hunk 1 above)
Comment (HarryMichal) [reply playbooks/shellcheck.yaml:8]: Ok, will do.

**Comment by HarryMichal** [comment]
> > Looks like I'll need to adjust the parts where the tests check what was written
> > in stdout/stderr. The spinner breaks those test cases.
> 
> Oh, really? Is it because the spinner just looks different now? I think that the spinner is written to the standard output in both the POSIX shell and Go versions.

I think the problem is in the `output` variable. I think it holds both `stdout` and `stderr`. I'll check the documentation.

File: playbooks/shellcheck.yaml (hunk 1 above)
Comment (debarshiray) [reply playbooks/shellcheck.yaml:8]: I took the liberty to split them out and merge them via https://github.com/containers/toolbox/pull/441 and https://github.com/containers/toolbox/pull/442

**Comment by HarryMichal** [comment]
@debarshiray This should be ready to be merged. The tests are not passing but they are working which is the main point of this PR. I'm just wondering about the additional dependency introduced in https://github.com/containers/toolbox/pull/438/commits/11133ac8bdd091eb4394ea6ab2bf3bd7065d42da. What do you think??

File: test/system/102-create.bats (hunk 2 above)
Comment (HarryMichal) [reply test/system/102-create.bats:10]: I updated the test cases to use both ways of creating a named container.

**Comment by HarryMichal** [comment]
Tests should be ok now. https://github.com/containers/toolbox/issues/410 still needs to be sorted out but that is a different issue (I suspect this is something introduced by Podman but I'm not sure. We'll see if V2 makes a difference regarding this matter). Rawhide tests are failing but that is due to an old version of Podman being available. Newer versions should have all the necessarry fixes.

**Comment by HarryMichal** [comment]
recheck

**Comment by debarshiray** [comment]
Thanks for updating the branch, @HarryMichal !

It seems that you switched from `golang.org/x/crypto/ssh/terminal` to `github.com/mattn/go-isatty`. I don't really have an opinion, but I am curious about what made you change your mind?

**Comment by debarshiray** [comment]
I split out and [merged](https://github.com/containers/toolbox/pull/495) the patch to drop the tests for Fedora 30 to keep things moving.

File: src/cmd/create.go [hunk 3]
```diff
@@ -181,6 +182,15 @@ func createContainer(container, image, release string, showCommandToEnter bool)
 		return nil
 	}
 
+	s := spinner.New(spinner.CharSets[9], 500*time.Millisecond)
+
+	if logLevel := logrus.GetLevel(); logLevel < logrus.DebugLevel && isatty.IsTerminal(1) {
```
Comment (debarshiray) [inline src/cmd/create.go:187]: I am curious. Why are we moving the entire chunk of code that creates the spinner, instead of just adding the extra condition to the existing chunk in-place?

**Comment by debarshiray** [comment]
> It seems that you switched from `golang.org/x/crypto/ssh/terminal` to
> `github.com/mattn/go-isatty`. I don't really have an opinion, but I am
> curious about what made you change your mind?

At first glance, anything that's under `golang.org/x` seems like a better choice because of its semi-official nature.

Looking at the API documentation of [golang.org/x/crypto/ssh/terminal](https://pkg.go.dev/golang.org/x/crypto/ssh/terminal) and [github.com/mattn/go-isatty](https://pkg.go.dev/github.com/mattn/go-isatty), it seems likely that the former will have a bigger footprint in terms of the final binary size. Preventing the binary size from growing is a worthwhile goal.

Podman uses `golang.org/x/crypto/ssh/terminal`, but one of its [dependencies](https://pkg.go.dev/github.com/vbauerster/mpb) uses `github.com/mattn/go-isatty`.

I see that `github.com/mattn/go-isatty` v0.0.12 is packaged in Fedora. However, strangely, it insists on a newer `golang.org/x/sys`, v0.0.0-20200116001909-b77594299b42, while Fedora only has `20200103gitc96a22e` which seems like a 13 days older snapshot.

**Comment by debarshiray** [comment]
> Looking at the API documentation of [golang.org/x/crypto/ssh/terminal](https://pkg.go.dev/golang.org/x/crypto/ssh/terminal)
> and [github.com/mattn/go-isatty](https://pkg.go.dev/github.com/mattn/go-isatty), it seems likely that the former will have a
> bigger footprint in terms of the final binary size. Preventing the
> binary size from growing is a worthwhile goal.

After actually trying out both packages and building the resulting binaries, `golang.org/x/crypto/ssh/terminal` gave the smaller size. 7778323 bytes versus 7782284.

It's not what I expected because `github.com/mattn/go-isatty` seemed like a much smaller package in comparison. It's essentially just:
```
import "golang.org/x/sys/unix"

// IsTerminal return true if the file descriptor is terminal.
func IsTerminal(fd uintptr) bool {
        _, err := unix.IoctlGetTermios(int(fd), unix.TCGETS)
        return err == nil
}
```

But I cannot argue with actual data. :) Unless I am missing something.

**Comment by debarshiray** [comment]
I discussed the spinner changes with @HarryMichal on Telegram, and he confirmed that there's no reason not to use the `golang.org/x/crypto/ssh/terminal` package. So, I took the liberty to tweak his commit accordingly, and merged it via https://github.com/containers/toolbox/pull/496

File: src/cmd/create.go [hunk 4]
```diff
@@ -634,10 +635,9 @@ func pullImage(image, release string) (bool, error) {
 
 	logrus.Debugf("Pulling image %s", imageFull)
 
-	if logLevel := logrus.GetLevel(); logLevel < logrus.DebugLevel {
+	if logLevel := logrus.GetLevel(); logLevel < logrus.DebugLevel && isatty.IsTerminal(1) {
 		s := spinner.New(spinner.CharSets[9], 500*time.Millisecond)
 		s.Prefix = fmt.Sprintf("Pulling %s: ", imageFull)
-		s.Writer = os.Stdout
```
Comment (debarshiray) [inline src/cmd/create.go:640]: Is there a reason to remove this line?

Comment (debarshiray) [reply src/cmd/create.go:640]: After a few rounds of testing it seems accidental to me, so I filed https://github.com/containers/toolbox/pull/498 to put it back.

But let me know if I am missing something. :)

File: src/cmd/rm.go [hunk 5]
```diff
@@ -70,6 +70,8 @@ func rm(cmd *cobra.Command, args []string) error {
 		return nil
 	}
 
+	var encounteredError bool = false
```
Comment (debarshiray) [inline src/cmd/rm.go:73]: Yes, I had noticed this regression while reviewing the Go rewrite in https://github.com/containers/toolbox/pull/318 I chose to ignore it because I couldn't think of a nice way to do it back then and didn't realize that it would have a negative effect on the tests.

It is a bit ugly to have each command directly call `os.Exit` this way. It's generally nicer to have a single return or exit point. In this case, we'd somehow propagate the error (and/or the exit code) up the function call stack, and have `Execute` in `cmd/root.go` handle it accordingly. However, I don't know off-hand how to achieve that with Cobra, but it should be possible.

In the worst case, given that, as far as we know, nobody is using `/usr/bin/toolbox` as an API from another program, I'd be happier to tweak the test suite to not expect a non-zero exit code, as a short-term temporary measure. Doesn't the test suite already look for the expected error message strings? If so, then I can live with it, while we figure out something nicer.

Comment (HarryMichal) [reply src/cmd/rm.go:73]: Fair enough. I'll drop those two commits and adjust the test cases.

**Comment by HarryMichal** [comment]
recheck

File: playbooks/fedora-31/setup-env.yaml [hunk 6]
```diff
@@ -5,6 +5,7 @@
       become: yes
       package:
         name:
+          - golang
```
Comment (debarshiray) [inline playbooks/fedora-31/setup-env.yaml:8]: It would be nicer to split this part out into a separate and smaller (eg., *add Go to the list of build requirements*) commit, instead of having it as part of a larger *update tests* commit.

File: test/system/301-rm.bats [hunk 7]
```diff
@@ -3,13 +3,13 @@
 load helpers
 
 @test "Try to remove a nonexistent container" {
-  run_toolbox 1 rm nonexistentcontainer
-  is "$output" "toolbox: failed to inspect $todelete" "Toolbox should fail with: no such container"
+  run_toolbox rm nonexistentcontainer
```
Comment (debarshiray) [inline test/system/301-rm.bats:6]: I'd probably merge the *Don't expect non-zero code on container removal* commit with the *update tests* commit, because that's what this is about - but still mention the reason in the commit message. :)

File: test/system/201-run.bats [hunk 8]
```diff
@@ -4,12 +4,12 @@ load helpers
 
 @test "Echo 'Hello World' inside of the default container" {
   run_toolbox run echo "Hello World"
-  is "$output" "Hello World" "Should say 'Hello World'"
+  # is "$output" "Hello World" "Should say 'Hello World'"
```
Comment (debarshiray) [inline test/system/201-run.bats:7]: Should we talk to the Zuul folks about this?

I am happy to turn off the assertion as a temporary measure. I just hope that we don't forget about it.

**Comment by debarshiray** [comment]
I [merged](https://github.com/containers/toolbox/pull/508) the renaming of the files and test names.

The only thing left is to figure out if we should notify Zuul about the *toolbox run* test failures.

File: test/system/201-run.bats (hunk 8 above)
Comment (HarryMichal) [reply test/system/201-run.bats:7]: A similar issue was brought up by the Zuul folks some time ago already (https://github.com/containers/toolbox/issues/410). I think we can track the issue there (the issue title should be updated, though).

Comment (debarshiray) [reply test/system/201-run.bats:7]: Ok, that's great.

Nitpick: we should link to https://github.com/containers/toolbox/issues/410 from this commit to complete the paper trail.

**Comment by debarshiray** [comment]
> The only thing left is to figure out if we should notify Zuul about the _toolbox run_ test failures.

Let's continue discussing that on https://github.com/containers/toolbox/pull/438

**Comment by juanje** [comment]
Hi, I was rewriting the bats tests (for practicing purposes) and I found this issue with the `toolbox run`.

Here the test (using [bats-assert](https://github.com/ztombol/bats-assert)):
```
@test "Echo 'Hello World' inside of the default container" {
  create_default_container
  run toolbox run echo "Hello World"

  assert_success
  assert_output "Hello World"
}
```

And the fail:
```
 ✗ Echo 'Hello World' inside of the default container
   (from function `assert_output' in file test/system/libs/bats-assert/src/assert.bash, line 239,
    in test file test/system/1201-run.bats, line 28)
     `assert_output "Hello World"' failed
   
   -- output differs --
   expected : Hello World
   actual   : Hello World
   --
```

As you can see, the outputs seem to be the same, but after some checking, I found that the `toolbox run echo` added a `CRLF` at the end of the line.

After much testing I tried this:
```
$ toolbox run echo foo > toolbox_output.txt
$ echo foo >  output.txt
$ file *output.txt
output.txt:         ASCII text
toolbox_output.txt: ASCII text, with CRLF line terminators
```

I couldn't make it work at the test comparison, but if I add the flag `-n` to the command `echo` the test work:
```
$ toolbox run echo -n foo > toolbox_output.txt
$ file toolbox_output.txt
toolbox_output.txt: ASCII text, with no line terminators
```

Well, I could make the comparison works with the assert's flag `--partial`, but it looks to me as a hack for the test to pass rather than a solution:
```
assert_output --partial "Hello World"
```

The same with the `echo -n` trick. The problem remains because if I try other commands it doesn't work. For example:
```
$ toolbox run ls > toolbox_output.txt
$ file toolbox_output.txt
toolbox_output.txt: ASCII text, with CRLF line terminators
```

I guess it's something about the toolbox output. I hope this give you some useful info to debug the problem.
